kong/plugins/opentelemetry/otlp.lua

```diff
@@ -0,0 +1,178 @@
+  for k, v in pairs(attr) do
+    local typ = type(v)
+    local pb_val
+
+    if typ == "string" then
+      pb_val = { string_value = v }
+
+    elseif typ == "number" then
+      pb_val = { double_value = v }
+
+    elseif typ == "boolean" then
+      pb_val = { bool_value = v }
+
+    else
+      pb_val = EMPTY_TAB -- considered empty
+    end
+
+    insert(pb_attributes, {
+      key = k,
+      value = pb_val,
+    })
+  end
+
+  return pb_attributes
+end
+
+local function transform_events(events)
+  if type(events) ~= "table" then
+    return nil
+  end
+
+  local pb_events = new_tab(#events, 0)
+  for _, evt in ipairs(events) do
+    local pb_evt = {
+      name = evt.name,
+      time_unix_nano = evt.time_ns,
+      -- dropped_attributes_count = 0,
+    }
+
+    if evt.attributes then
+      pb_evt.attributes = transform_attributes(evt.attributes)
+    end
+
+    insert(pb_events, pb_evt)
+  end
+
+  return pb_events
+end
+
+local function transform_span(span)
+  assert(type(span) == "table")
+
+  local pb_span = {
+    trace_id = span.trace_id,
+    span_id = span.span_id,
+    -- trace_state = "",
+    parent_span_id = span.parent_id or "",
+    name = span.name,
+    kind = span.kind or 0,
+    start_time_unix_nano = span.start_time_ns,
+    end_time_unix_nano = span.end_time_ns,
+    attributes = span.attributes and transform_attributes(span.attributes),
+    -- dropped_attributes_count = 0,
+    events = span.events and transform_events(span.events),
+    -- dropped_events_count = 0,
+    -- links = EMPTY_TAB,
+    -- dropped_links_count = 0,
+    status = span.status and PB_STATUS[span.status],
+  }
+
+  return pb_span
+end
+
+local function to_pb(data)
+  return pb.encode("opentelemetry.proto.collector.trace.v1.ExportTraceServiceRequest", data)
+end
+
+local encode_traces
+do
+  local attributes_cache = setmetatable({}, { __mode = "k" })
+  local function default_resource_attributes()
+    return {
+      ["service.name"] = "kong",
+      ["service.instance.id"] = kong and kong.node.get_id(),
+      ["service.version"] = kong and kong.version,
+    }
+  end
+
+  local function render_resource_attributes(attributes)
+    attributes = attributes or EMPTY_TAB
+
+    local resource_attributes = attributes_cache[attributes]
+    if resource_attributes then
+      return resource_attributes
+    end
+
+    local default_attributes = default_resource_attributes()
+    resource_attributes = table_merge(default_attributes, attributes)
+
+    resource_attributes = transform_attributes(resource_attributes)
+    attributes_cache[attributes] = resource_attributes
+
+    return resource_attributes
+  end
+
+  local pb_memo = {
+    resource_spans = {
+      {
+        resource = {
+          attributes = {}
+        },
+        scope_spans = {
+          {
+            scope = {
+              name = "kong-internal",
+              version = "0.1.0",
+            },
+            spans = {},
+          }
```

Comment:
Should we add `,` after each `}`

Reply:
I wonder if it is an option that could be in _luacheck_ for all the code style-related.
Thanks, anyway I will solve it now.